core/src/main/java/io/confluent/rest/RateLimitNetworkTrafficListener.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2023 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.confluent.rest;
+
+import io.confluent.rest.ratelimit.NetworkTrafficRateLimiter;
+import io.confluent.rest.ratelimit.NetworkTrafficRateLimiterFactory;
+import java.net.Socket;
+import java.nio.ByteBuffer;
+import org.eclipse.jetty.io.NetworkTrafficListener;
+
+public class RateLimitNetworkTrafficListener implements NetworkTrafficListener {
+  private final NetworkTrafficRateLimiter rateLimiter;
+
+  public RateLimitNetworkTrafficListener(RestConfig restConfig) {
+    rateLimiter = NetworkTrafficRateLimiterFactory.create(restConfig);
+  }
+
+  @Override
+  public void incoming(final Socket socket, final ByteBuffer bytes) {
+    int cost = bytes.limit() - bytes.position();
```

Comment:
Also, would there be scenarios where operationally would want to know if back-pressure is being applied?

Worthwhile to add `log.trace` for now for future debugging, if rateLimiter.rateLimit() > 0 i.e. it had to block to get permits.

Reply:
For sure, I can add a debug line (to avoid too verbose in prod)